### PR TITLE
IGNITE-13543 Calcite integration. Sort-based aggregates

### PR DESCRIPTION
**Pull request by korlov42**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+                if (cmp == 0)
+                    grp.add(row);
+                else {
+                    if (cmpRes == 0)
+                        cmpRes = cmp;
+                    else
+                        assert cmp == cmpRes : "Input not sorted";
+
+                    doPush();
```
Comment (korlov42) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java:130]: right now it's possible to push more row than actually requested. Please see my PR with test that highlight this issue: https://github.com/gridgain/apache-ignite/pull/262 (this includes some minor fixes)

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+                    grp = newGroup(row);
+                }
+            }
+            else
+                grp = newGroup(row);
+
+            prevRow = row;
+
+            if (waiting == 0)
+                source().request(waiting = IN_BUFFER_SIZE);
```
Comment (korlov42) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java:141]: it's better to request new batch in another task to allow other queries do progress

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+            waiting = -1;
+
+            if (grp != null)
+                doPush();
+
+            grp = null;
+            prevRow = null;
+
+            downstream().end();
```
Comment (korlov42) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java:164]: downstream's `end` should not be called subsequently with current `end()`

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/SortAggregateConverterRule.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override protected PhysicalNode convert(RelOptPlanner planner, RelMetadataQuery mq,
+        LogicalAggregate rel) {
+        RelOptCluster cluster = rel.getCluster();
+        RelTraitSet inTrait = cluster.traitSetOf(IgniteConvention.INSTANCE);
+        RelTraitSet outTrait = cluster.traitSetOf(IgniteConvention.INSTANCE);
+        RelNode input = convert(rel.getInput(), inTrait);
+        if (F.isEmpty(rel.getGroupSet()))
```
Comment (zstan) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/SortAggregateConverterRule.java:54]: is it correct here ?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMapAggregateSort.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        RelNode input,
+        ImmutableBitSet groupSet,
+        List<ImmutableBitSet> groupSets,
+        List<AggregateCall> aggCalls,
+        RelCollation collation
+    ) {
+        super(cluster, traitSet, input, groupSet, groupSets, aggCalls);
+
+        assert Objects.nonNull(collation);
+        assert !collation.isDefault();
```
Comment (aimiebell) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMapAggregateSort.java:64]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Use of identifier that is a keyword in later versions of Java (NM_FUTURE_KEYWORD_USED_AS_IDENTIFIER)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-use-of-identifier-that-is-a-keyword-in-later-versions-of-java-nm-future-keyword-used-as-identifier).
The identifier is a word that is reserved as a keyword in later versions of Java, and your code will need to be changed in order to compile it in later versions of Java.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/LogicalRelImplementor.java [hunk 6]
```diff
@@ -427,8 +433,8 @@ public LogicalRelImplementor(
     }
 
     /** {@inheritDoc} */
-    @Override public Node<Row> visit(IgniteAggregate rel) {
-        AggregateNode.AggregateType type = AggregateNode.AggregateType.SINGLE;
+    @Override public Node<Row> visit(IgniteAggregateHash rel) {
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/LogicalRelImplementor.java:436]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        if (idx != 0)
+            throw new IndexOutOfBoundsException();
+
+        return this;
+    }
+
+    /** */
+    private Group newGroup(Row r) {
+        if (type == AggregateType.REDUCE)
+            System.out.println();
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java:190]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregateHash.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        return new IgniteReduceAggregateHash(getCluster(), traits, input, groupSet, groupSets, aggCalls, getRowType());
+    }
+
+    /** {@inheritDoc} */
+    @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        double rows = mq.getRowCount(getInput());
+
+        // TODO: fix it when https://issues.apache.org/jira/browse/IGNITE-13543 will be resolved
+        // currently it's OK to have such a dummy cost because there is no other options
+        return planner.getCostFactory().makeCost(rows, rows * IgniteCost.ROW_PASS_THROUGH_COST, 0);
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregateHash.java:110]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregateSort.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    /** {@inheritDoc} */
+    @Override public RelWriter explainTerms(RelWriter pw) {
+        return super.explainTerms(pw).item("collation", collation);
+    }
+
+    /** {@inheritDoc} */
+    @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughCollation(RelTraitSet nodeTraits, List<RelTraitSet> inputTraits) {
+        RelCollation collation = RelCollations.of(ImmutableIntList.copyOf(groupSet.asList()));
+
+        return Pair.of(nodeTraits.replace(RelCollations.EMPTY),
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregateSort.java:97]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMapAggregateHash.java [hunk 10]
```diff
@@ -113,8 +93,11 @@ public static RelDataType rowType(RelDataTypeFactory typeFactory) {
...
     /** {@inheritDoc} */
-    @Override public IgniteRel clone(RelOptCluster cluster, List<IgniteRel> inputs) {
-        return new IgniteMapAggregate(cluster, getTraitSet(), sole(inputs),
-            getGroupSet(), getGroupSets(), getAggCallList());
+    @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        double rows = mq.getRowCount(getInput());
+
+        // TODO: fix it when https://issues.apache.org/jira/browse/IGNITE-13543 will be resolved
+        // currently it's OK to have such a dummy cost because there is no other options
+        return planner.getCostFactory().makeCost(rows, rows * IgniteCost.ROW_PASS_THROUGH_COST, 0);
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMapAggregateHash.java]: Fixed

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/BaseAggregateTest.java [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+@WithSystemProperty(key = "calcite.debug", value = "true")
+public abstract class BaseAggregateTest extends AbstractExecutionTest {
+    /** Last parameter number. */
+    protected static final int TEST_AGG_PARAM_NUM = LAST_PARAM_NUM + 1;
+
+    /** */
+    @Parameterized.Parameter(TEST_AGG_PARAM_NUM)
+    public TestAggregateType testAgg;
+
+    @Parameterized.Parameters(name = PARAMS_STRING + ", type={" + TEST_AGG_PARAM_NUM + "}")
```
Comment (AMashenkov) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/BaseAggregateTest.java:66]: Javadoc

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteSortAggregate.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughCollation(RelTraitSet nodeTraits, List<RelTraitSet> inputTraits) {
+        RelCollation collation = RelCollations.of(ImmutableIntList.copyOf(groupSet.asList()));
+
+        return Pair.of(nodeTraits.replace(collation),
+            ImmutableList.of(inputTraits.get(0).replace(collation)));
+    }
+
+    /** {@inheritDoc} */
+    @Override public List<Pair<RelTraitSet, List<RelTraitSet>>> deriveCollation(RelTraitSet nodeTraits, List<RelTraitSet> inputTraits) {
+        RelCollation requiredCollation = RelCollations.of(ImmutableIntList.copyOf(groupSet.asList()));
```
Comment (korlov42) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteSortAggregate.java:108]: actually, any permutation of group columns is possible here

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortAggregateNode.java [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+
+        try {
+            checkState();
+
+            requested = rowsCnt;
+
+            if (waiting == 0) {
+                waiting = requested;
+
+                source().request(requested);
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortAggregateNode.java:104]: In this case we have to use buffer in the node.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortAggregateNode.java [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+        Supplier<List<AccumulatorWrapper<Row>>> accFactory,
+        RowFactory<Row> rowFactory,
+        Comparator<Row> comp
+    ) {
+        super(ctx, rowType);
+
+        this.type = type;
+        this.accFactory = accFactory;
+        this.rowFactory = rowFactory;
+        this.grpSet = grpSet;
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortAggregateNode.java:86]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregate.java [hunk 15]
```diff
@@ -17,258 +17,112 @@
...
-                }
-
-                break;
-
-            default:
-                break;
+    /** */
+    public double estimateMemoryForGroup(RelMetadataQuery mq) {
+        if (aggCalls.isEmpty())
+            return groupSet.cardinality() * IgniteCost.AVERAGE_FIELD_SIZE;
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregate.java:77]: changed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMapAggregateBase.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.AggregateCall;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import static org.apache.ignite.internal.processors.query.calcite.trait.TraitUtils.changeTraits;
+
+/**
+ *
+ */
+public abstract class IgniteMapAggregateBase extends IgniteAggregate implements IgniteRel {
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMapAggregateBase.java]: removed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdDistinctRowCount.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    /** {@inheritDoc} */
+    @Override public Double getDistinctRowCount(
+        RelNode rel,
+        RelMetadataQuery mq,
+        ImmutableBitSet groupKey,
+        RexNode predicate
+    ) {
+        double rowCount = mq.getRowCount(rel);
+
+        rowCount *= 1.0 - Math.pow(.5, groupKey.cardinality());
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdDistinctRowCount.java:48]: fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregate.java [hunk 18]
```diff
@@ -17,258 +17,112 @@
...
+        double rows = mq.getRowCount(getInput());
 
-            assert diff.size() == 1 && F.first(diff) == TraitUtils.distribution(outTraits);
-        }
+        double groupsCnt = estimateRowCount(mq);
 
-        RelNode map = new IgniteMapAggregate(getCluster(), in, convert(getInput(), in), groupSet, groupSets, aggCalls);
-        return new IgniteReduceAggregate(getCluster(), outTraits, convert(map, outTraits), groupSet, groupSets, aggCalls, getRowType());
+        return costFactory.makeCost(
+            groupsCnt,
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregate.java:92]: fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregate.java [hunk 19]
```diff
@@ -17,258 +17,112 @@
...
-    @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
         double rows = mq.getRowCount(getInput());
 
-        // TODO: fix it when https://issues.apache.org/jira/browse/IGNITE-13543 will be resolved
-        // currently it's OK to have such a dummy cost because there is no other options
-        return planner.getCostFactory().makeCost(rows, rows * IgniteCost.ROW_PASS_THROUGH_COST, 0);
+        double groupsCnt = estimateRowCount(mq);
+
+        return costFactory.makeCost(
+            groupsCnt,
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteAggregate.java:109]: fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/SortAggregateConverterRule.java (hunk 4 above)
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/SortAggregateConverterRule.java]: Moved to the top of the method

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/BaseAggregateTest.java (hunk 11 above)
Comment (tledkov-gridgain) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/BaseAggregateTest.java:66]: fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortAggregateNode.java [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+        this.comp = comp;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void request(int rowsCnt) {
+        assert !F.isEmpty(sources()) && sources().size() == 1;
+        assert rowsCnt > 0 && requested == 0;
+        assert waiting <= 0;
+
+        try {
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortAggregateNode.java:96]: fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortAggregateNode.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortAggregateNode.java]: fixed

**Pull request closed without merging**